torchaudio/compliance/kaldi.py

```diff
@@ -512,3 +514,172 @@ def fbank(
+
+    t_eq_zero_indices = delta_t.eq(0.0)
+    t_not_eq_zero_indices = ~t_eq_zero_indices
+    weights[t_not_eq_zero_indices] *= torch.sin(
+        2 * math.pi * lowpass_cutoff * delta_t[t_not_eq_zero_indices]) / (math.pi * delta_t[t_not_eq_zero_indices])
+    # limit of the function at t = 0
+    weights[t_eq_zero_indices] *= 2 * lowpass_cutoff
+
+    weights /= orig_freq  # size (output_samples_in_unit, max_weight_width)
+    return min_input_index, weights
+
+
+def _lcm(a, b):
+    return abs(a * b) // math.gcd(a, b)
+
+
+def _get_num_LR_output_samples(input_num_samp, samp_rate_in, samp_rate_out):
+    """ Based on LinearResample::GetNumOutputSamples
+    Inputs:
+        input_num_samp (int): the number of samples in the input
+        samp_rate_in (float): the original frequency of the signal
+        samp_rate_out (float): the desired frequency
+
+    Outputs:
+        int: the number of output samples
+    """
+    # For exact computation, we measure time in "ticks" of 1.0 / tick_freq,
+    # where tick_freq is the least common multiple of samp_rate_in and
+    # samp_rate_out.
+    tick_freq = _lcm(samp_rate_in, samp_rate_out)
+    ticks_per_input_period = tick_freq // samp_rate_in
+
+    # work out the number of ticks in the time interval
+    # [ 0, input_num_samp/samp_rate_in ).
+    interval_length_in_ticks = input_num_samp * ticks_per_input_period
+    if interval_length_in_ticks <= 0:
+        return 0
+    ticks_per_output_period = tick_freq // samp_rate_out
+    # Get the last output-sample in the closed interval, i.e. replacing [ ) with
+    # [ ].  Note: integer division rounds down.  See
+    # http://en.wikipedia.org/wiki/Interval_(mathematics) for an explanation of
+    # the notation.
+    last_output_samp = interval_length_in_ticks // ticks_per_output_period
+    # We need the last output-sample in the open interval, so if it takes us to
+    # the end of the interval exactly, subtract one.
+    if last_output_samp * ticks_per_output_period == interval_length_in_ticks:
+        last_output_samp -= 1
+    # First output-sample index is zero, so the number of output samples
+    # is the last output-sample plus one.
+    num_output_samp = last_output_samp + 1
+    return num_output_samp
+
+
+def resample_waveform(wave, orig_freq, new_freq):
+    """Resamples the wave at the new frequency. This matches Kaldi's OfflineFeatureTpl ResampleWaveform
+    which uses a LinearResample (resample a signal at linearly spaced intervals to upsample/downsample
+    a signal).
+
+    Inputs:
+        wave (Tensor): the input signal of size (c, n)
+        orig_freq (float): the original frequency of the signal
+        new_freq (float): the desired frequency
+
+    Outputs:
+        Tensor: the signal at the new frequency
+    """
+    assert orig_freq > 0.0 and new_freq > 0.0
+
+    min_freq = min(orig_freq, new_freq)
+    lowpass_cutoff = 0.99 * 0.5 * min_freq
```

Comment:
What is the meaning/purpose of the 0.99 here?

Reply:
```
The filter cutoff needs to be less than samp_rate_in_hz/2 
and less than samp_rate_out_hz/2. 

Assume S is the original  sampling frequency. In general 
we want to bandlimit to lower than S/2, because we don't have a perfect filter 
and also  because if we want to resample at a lower frequency than S, 
we need to bandlimit to below half of that. 
```

Reply:
Ok, you can add a comment that details this rational here.